### PR TITLE
1091/mk/void withdrawal

### PR DESCRIPTION
**Pull request by mkurapov**
## Changes proposed in this pull request

- Implements `voidAccountWithdrawal`  & `postAccountWithdrawal` functions

## Context

- Progress towards #1091 

## Checklist

- [x] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass

### CODE REVIEW

File: packages/backend/src/accounting/psql/service.ts [hunk 1]
```diff
@@ -286,9 +293,9 @@ async function createAccountWithdrawal(
 
 async function voidAccountWithdrawal(
   deps: ServiceDependencies,
-  withdrawalId: string
+  transferRef: string
 ): Promise<void | TransferError> {
-  throw new Error('Not implemented')
+  return await voidTransfer(deps, transferRef)
```
Comment (mkurapov) [inline packages/backend/src/accounting/psql/service.ts:298]: could probably just have the service call directly, but kept it for consistency?

File: packages/backend/src/accounting/psql/ledger-transfer/index.test.ts [hunk 2]
```diff
@@ -513,4 +520,193 @@ describe('Ledger Transfer', (): void => {
...
+      const trx = await knex.transaction()
+
+      await expect(
+        getTransferForUpdate(serviceDeps, transfer.transferRef, trx)
+      ).resolves.toEqual(transfer)
+
+      const timeoutOrTransfer = await Promise.race([
+        transfer.$query().patchAndFetch({ state: LedgerTransferState.PENDING }),
+        new Promise((reject) => setTimeout(() => reject('timeout'), 5000))
+      ])
```
Comment (mkurapov) [inline packages/backend/src/accounting/psql/ledger-transfer/index.test.ts:575]: could definitely decrease the time, but I thought this was an ok method to test a lock

File: packages/backend/src/accounting/psql/ledger-transfer/index.ts [hunk 3]
```diff
@@ -75,6 +75,43 @@ export async function getAccountTransfers(
   )
 }
 
+export async function getTransferForUpdate(
+  _: ServiceDependencies,
+  transferRef: string,
+  trx: Transaction
```
Comment (mkurapov) [inline packages/backend/src/accounting/psql/ledger-transfer/index.ts:81]: I think I'll have `createTransfers` also take in a `Transaction` directly

File: packages/backend/src/accounting/psql/ledger-transfer/model.ts [hunk 4]
```diff
@@ -41,6 +41,22 @@ export class LedgerTransfer extends BaseModel {
...
+    return this.state === LedgerTransferState.POSTED
+  }
+
+  public get isVoided(): boolean {
+    return this.state === LedgerTransferState.VOIDED
+  }
+
+  public get expired(): boolean {
+    return this.isPending && !!this.expiresAt && this.expiresAt <= new Date()
+  }
```
Comment (mkurapov) [inline packages/backend/src/accounting/psql/ledger-transfer/model.ts:58]: will add tests for these in the AM

**Comment by mkurapov** [comment]
@sabineschaller actually going to combine this with #1131, it'll be easier to review the `voidWithdrawal` and `postWithdrawal` together

File: packages/backend/package.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "backend",
   "scripts": {
-    "test": "jest --passWithNoTests --maxWorkers=50%",
+    "test": "jest --passWithNoTests --maxWorkers=50% --logHeapUsage",
```
Comment (wilsonianb) [inline packages/backend/package.json:4]: :eyes: 
Is this looking into the inconsistent `backend` test failures?
Does `--logHeapUsage` this need `--runInBand`?
https://jestjs.io/docs/cli#--logheapusage

File: packages/backend/src/accounting/psql/ledger-transfer/index.test.ts [hunk 6]
```diff
@@ -513,4 +519,146 @@ describe('Ledger Transfer', (): void => {
...
+          debitAccountId: settlementAccount.id,
+          amount: 10n,
+          ledger: account.ledger,
+          state: LedgerTransferState.PENDING
+        },
+        knex
+      )
+    })
+
+    test('posts transfer', async (): Promise<void> => {
```
Comment (wilsonianb) [inline packages/backend/src/accounting/psql/ledger-transfer/index.test.ts:539]:
```suggestion
    test('voids transfer', async (): Promise<void> => {
```

File: packages/backend/src/accounting/psql/ledger-transfer/index.test.ts [hunk 7]
```diff
@@ -513,4 +519,146 @@ describe('Ledger Transfer', (): void => {
...
+      ).resolves.toEqual(TransferError.AlreadyPosted)
+    })
+
+    test('returns error if transfer already voided', async (): Promise<void> => {
+      await expect(
+        voidTransfer(serviceDeps, transfer.transferRef)
+      ).resolves.toBeUndefined()
+
+      await expect(
+        voidTransfer(serviceDeps, transfer.transferRef)
```
Comment (wilsonianb) [inline packages/backend/src/accounting/psql/ledger-transfer/index.test.ts:660]: -> `postTransfer`

File: packages/backend/src/accounting/psql/ledger-transfer/model.ts [hunk 8]
```diff
@@ -41,6 +41,22 @@ export class LedgerTransfer extends BaseModel {
...
+
+  public get isPosted(): boolean {
+    return this.state === LedgerTransferState.POSTED
+  }
+
+  public get isVoided(): boolean {
+    return this.state === LedgerTransferState.VOIDED
+  }
+
+  public get expired(): boolean {
```
Comment (wilsonianb) [inline packages/backend/src/accounting/psql/ledger-transfer/model.ts:56]:
```suggestion
  public get isExpired(): boolean {
```
for consistency

File: packages/backend/src/accounting/psql/ledger-transfer/model.ts [hunk 9]
```diff
@@ -41,6 +41,22 @@ export class LedgerTransfer extends BaseModel {
     this.transferRef = this.transferRef || uuid()
   }
 
+  public get isPending(): boolean {
+    return this.state === LedgerTransferState.PENDING
+  }
```
Comment (wilsonianb) [inline packages/backend/src/accounting/psql/ledger-transfer/model.ts:46]: :thinking: should this return false if expired?

File: packages/backend/package.json (hunk 5 above)
Comment (wilsonianb) [reply packages/backend/package.json:4]: oh
- https://github.com/interledger/rafiki/pull/1132

Comment (mkurapov) [reply packages/backend/package.json:4]: Yep, we had an out-of-memory exception earlier, so wanted to debug the tests

File: packages/backend/src/accounting/psql/ledger-transfer/model.ts (hunk 9 above)
Comment (mkurapov) [reply packages/backend/src/accounting/psql/ledger-transfer/model.ts:46]: Hmm, I might just remove this method and do
```ts
  public get expired(): boolean {
    return (
      this.state === LedgerTransferState.PENDING &&
      !!this.expiresAt &&
      this.expiresAt <= new Date()
    )
  } 
```

**Review by wilsonianb — APPROVED** [review]

**Pull request merged**